File: ports/java/src/core/containers/BcsMap.java
package bcs.core.containers;

import java.util.concurrent.ConcurrentHashMap;
import java.util.Optional;

// BcsMap maps to ConcurrentHashMap for thread-safe key-value storage
public class BcsMap<K, V> {
    private final ConcurrentHashMap<K, V> items;

    public BcsMap() {
        this.items = new ConcurrentHashMap<>();
    }

    public void insert(K key, V value) {
        this.items.put(key, value);
    }

    public Optional<V> value(K key) {
        return Optional.ofNullable(this.items.get(key));
    }

    public boolean contains(K key) {
        return this.items.containsKey(key);
    }
}

File: ports/csharp/src/core/containers/BcsMap.cs
using System.Collections.Concurrent;

namespace BCS.Core.Containers {
    // BcsMap maps to ConcurrentDictionary for thread-safe key-value storage
    public class BcsMap<K, V> {
        private ConcurrentDictionary<K, V> _items;

        public BcsMap() {
            _items = new ConcurrentDictionary<K, V>();
        }

        public void Insert(K key, V value) {
            _items[key] = value;
        }

        public bool TryGetValue(K key, out V value) {
            return _items.TryGetValue(key, out value);
        }

        public bool Contains(K key) {
            return _items.ContainsKey(key);
        }
    }
}
